\documentclass[11pt]{article}

% //Spec: Use helvetica font (similar to Arial)
\renewcommand{\familydefault}{\sfdefault}
\usepackage[scaled=1]{helvet}
\usepackage[helvet]{sfmath}

\usepackage{setspace}  % Spacing between lines
\usepackage{amsmath}  % math equation packages
\usepackage{graphicx}  % To be able to insert images
\everymath={\sf}  % Use helvetica to print equations too

% //W: Kept because error if these commands are removed
\title{}
\date{}
\author{}


\begin{document}
  \begin{titlepage}
    \centering
    \includegraphics[width=0.5\textwidth]{imgs/logourv}  % Logo top
    \par        
    \vspace{1cm}
    \Large
    {Digenic and trigenic interaction prediction using Mixed-Membership Stochastic Blockmodels \par}
    \vspace{1cm}
    {Aleix Mariné-Tena \par}
    \vspace{1cm}
    {\itshape\Large BIOTECHNOLOGY FINAL PROJECT \par}
    \vfill

    \vspace{1cm} 
    \includegraphics[width=0.7\textwidth]{imgs/grafportada}  % Portada img
    \par    
    \vspace{1cm}  
    \vfill
    
    \large
    \raggedright
    {Tutor and supervisor: Marta Sales-Pardo, ETSEQ, (marta.sales@urv.cat) \par}
    {In cooperation with: SEES:lab \par}
 	\vspace{2cm}
    
    \raggedleft
    {\large September 2020 \par}
  \end{titlepage}
  
  \pagenumbering{gobble}
  \newpage

  % Index
  \pagenumbering{arabic}
  \tableofcontents
  \newpage

\setlength{\parskip}{1em}  % Set vertical separation between paragraphs
\onehalfspacing  % //Spec: spacing 1.5
\normalsize  % //Spec: normalsize = 11 pt (declared at e headers)

\section{Abstract}
\par
In this project we will implement a mathematical model based in Mixed-Membership Stochastic Blockmodels to be able to do predictions of the strength and type of genetic interaction between two or three genes from the human model organism \textit{Pichia pastoris}. 

\par
We will use a dataset of 501510 entries containing the fitness data from yeast triple and double \textit{knock-out}  mutants, each with a different combination of mutated genes.

\par
After validating the predictions of the model using different metrics, we will compare how genes related according to Mixed-Membership Stochastic Blockmodels are related in the Gene Onthology therms.

\par
\textbf{Keywords:} Mixed-membership stochastic blockmodels, machine learning, genetic interaction, Gene Onthology therms.


\section{Introduction}
The increasing amount of available information, the cheapening of computation power and the need for tools that are capable of digesting large amounts of data, have made machine learning one of the most efficient ways to do predictions from incomplete data.

\par 
Previous studies have shown the potential of Mixed-Membership Stochastic Blockmodels (MMSBM) to do accurate predictions from nonobserved data in a scalable algorithm. MMSBM is a machine-learning technique that can be applied to different kind of data to do accurate predictions if the training dataset is large enough. 

\par
In Biology, as in other science fields, available data in public databases is growing every year specially due to the advance in multiplexing techniques such as ELISA, DNA arrays or Luminex. That points out the increasing need of bioinformatics and machine-learning in biotechnology research in order to extract conclusion from these large amounts of data.

\par
In this study we put the focus on genetic interaction in the human model organism \textit{Pichia pastoris} (\textit{P. pastoris}). There are big datasets available of genetic interaction between two or three genes in this organism, but they just represent ~1\% of the possible genetic interactions in (\textit{P. pastoris}). This fact makes this problem ideal for a machine-learning solution because it will give predictions from genetic interaction without having to do it experimentally in the lab. 

\section{Objective}
  \begin{itemize}
    \item Develop an algorithm that is able to train from our dataset to do predictions from non-observed data.
    \item Compute different models with different parameters and obtain their performance metrics in order to select the best model.
    \item Compare the clustering of the genes in our best model with the distance between their respective gene onthology therms.
  \end{itemize}

\section{Theoretical Background}
  \subsection{Genetic Interaction}
    The central dogma of biochemistry states that the information in most living beings flows from genes, which are the fundamental elements of the genetic material, to proteins and RNA, which are the responsible for many heterogeneous functions in the cell: structural, reaction catalysis, signaling, transcription factors...
    
     \par 
    We can understand genetics following a "static" approach in which each gene is studied individually to determine its function and to determine what are the chemical and physical conditions needed in the cell for that particular gene to activate and give its final product (protein or RNA). 

    \par 
    But that is just a simplification from reality since Genetics has a dynamic and very complex behaviour, specially in superior organisms: Genes and their products are interacting with each other in many ways. Usually, this interactions are very subtile and have functions that have not been discovered yet. 
    
    \par 
    So, understanding the way in which genes and their products interact with each other can be a difficult task, since many involved factors are still unknown. But, experimentally, observing a chosen phenoype, we can determine if two or more genes are actually interacting, following the next definition:  
    
    \paragraph{Genetic Interaction} \textit{Genetic interaction is a phenomenon that occur when two or more mutant alleles in a single individual combine to result in a phenotype that is different from the expected phenotype when these alleles are tested separately in different individuals.}

    \par
    For example, let A and B be the only two genes present in \textit{P. pastoris} that code for an enzyme responsible of a step in a vital pathway. Also, let \textbf{fitness} be a numeric phenotype than can be calculated as the ratio between the real diameter of the colony and the diameter of the \textit{wild-type} colony. 
    
    \par
    When the gene A is non-functional or missing, gene B can replace gene A's function and vice versa. This process allows \textit{P. pastoris} to grow even when some vital genes are deleted. Consequently, the fitness phenotype when A or B are deleted in a individual should be ~1 or \textbf{non-lethal}.

    \par
    But if we delete gene A and B in the same individual, we will find that the fitness phenotype is 0 because \textit{P. pastoris} will not be able to grow up, because there is no gene this time that can replace the function of the lost vital genes.

    \par
    Knowing all the above we can realise that A and B are interacting because when deleted in the same individual the obtained phenotype is different from the expected phenotype: Since genes A and B have \textbf{non-lethal} phenotype when deleted separately, we expect the same when combined in the same individual. Note that this is an ilustrative example and genetic interaction is not always so drastic and clear.
    
    \par
    The study of the fitness phenotype allows to easily determine how lethal the supression of genes can be, and also gives us the possibility of defining a formula to calculate a parameter that expresses the difference between the observed and the expected fitness of a mutant colony. Given a certain threshold of this parameter we can affirm that there is genetic interaction and also we can determine what type of genetic interaction is happening. 

      \subsubsection{Types of genetic interaction}
      We will consider two main types of interaction:
      \begin{enumerate}
        \item \textbf{Negative genetic interaction:} Occurs when a combination of mutations leads to a fitness defect that is more exacerbated than expected. 
		\begin{itemize}
        \item \textbf{Synthetic lethality} occurs when two \textit{nonletal} mutations generate a \textit{lethal} mutant when combined.
        \end{itemize}

        \item \textbf{Positive genetic interaction:} Occurs when a combination of mutations leads to a fitness greater than expected. 
		\begin{itemize}
        \item \textbf{Genetic suppression}: Occurs when the mutations in the fitness defect of a query mutant is alleviated by a mutation in a second gene. 
        \end{itemize}
      \end{enumerate}




  
  \subsection{\textit{P. pastoris} as a human model organism}

  \subsection{Machine Learning}
  
  \subsection{Stochastic Blockmodels}
  
    \subsubsection{Mixed-Membership Stochastic Blockmodels}
    
  
  
  
Fonament Teóric: Aquí hi aniria un seguit de subtemes. Primer aniria una explicació del què és el machine learning i el model mixed-membership SBM en llenguatge natural. Després faria servir les equacions per a ilustrar com l'apliquem en aquest cas per a les dades que tenim.  També comentaria l'organisme d'on treiem els gens (Pichia pastoris) ja que es un model humà i per tant els resultats es poden extrapolar per a diversos processos celulars. També cal explicar el funcionament dels gene onthology therm i les classificacions que hi ha.

\section{Methodologies}

  \subsection{Materials}
  This section covers all the elements needed to do our experiment. Since our experiment is an algorithm its materials are basically hardware and software elements.
    \subsubsection{Hardware requirements}
    The training algorithm is the only algorithm that is strongly hardware-demanding. Using just one thread of execution in a last generation user PC, the algorithm can delay at most two weeks to finish. That is why there is an extra layer of parallelization that allows to run the algorithm in parallel using the number of threads corresponding to the number of cores available in the machine running the algorithm. That allows to divide the execution time by the number of cores of the machine running the algorithm, so better performances are expected in better machines with more cores.
    \subsubsection{Software Requirements}
    Due to having big datasets, long scripts and the need for computing power, many software dependencies need to be satisfied before working with the algorithm. Some are completely necessary and some others are just accessories for comodity and data visualization.
    \par 
    Anyway, since there are many tools to download and configure a side project
      \paragraph{Python3}     
      The main algorithm is written in Python3 so it needs a Python3 interpreter in order to run the code. The algorithm also uses the following Python3 packages that need to be installed:
        \begin{itemize}
          \item NumPy: Used for random picking function.
          \item codecs: To read from different files.
          \item matplotlib: To create all the graphics from the results.
        \end{itemize}
      \paragraph{PyPy3}
      PyPy3 is a Python3 interpreter that runs much faster in many cases. We strongly recommend to use this interpreter since it reduces the computation time of the training algorithm.
      \paragraph{Linux}
      Even though Python scripts are portable, we strongly recommend to use Linux to run the script. The last version of the code is tested in Ubuntu 20.04 LTS Focal.
      \paragraph{GNU-parallel}
      GNU-parallel is a Linux utility that allows to easily run batches of commands in parallel. It is not needed to run the main algorithm, but is needed to run the parallelization layer, which is strongly recommended to do.
      \paragraph{Git}
      Git is a software versioning tool to maintain and develop software. Git is needed at least to clone the repository where all the datasets and scripts are stored. 
      \par
      Also, because of the need to store all the datasets in the same repository, the complementary utility Git-LFS was installed in the repository. This utility allows to work with big datasets in the repository without sacrificing speed when cloning and pulling.
      \par 
      A .gitignore file was defined for all the possible junk, temporals and product of compilation for our repository in order to ignore all those files when updating the repository.
      \paragraph{LaTeX}
      The documentation was written in LaTeX so a LaTeX compilator is needed to obtain the final document of the documentation.

  \subsection{Design}
  Disseny: Esquema de funcionament del software amb respecte a la especificació.

  \subsection{Implementation}
  Implementació: Funcionament del script principal (TrigenicInteractionPredictor\_23.py) a mig detall, a nivell del main, i de l'script que s'encarrega de fer el processament dels GO therm.

\section{Results and Discussion}
Resultats i discussió: Dades obtingudes al final de l'experiment i raonament de quines idees podem extreure d'aquestes dades. Aquí aniria el gràfic de AUC vs held-outlikelihood amb les dades agrupades segons k i la justificació de la k que triem com a millor. També anirien els resultats que extreurem de la part dels GO therm.

\section{Conclusions}
Conclusions del treball

\section{Bibliography}
http://www.urv.cat/ca/vida-campus/serveis/crai/que-us-oferim/suport-investigacio/citacio/

\section{Autoevaluation}

\section{Annex}

\end{document}
